wordpress-mobile/WordPress-Android
# Start uploadService as foreground service

Fixes #9982


The issue basically is that we are sending tons of non-fatal exception to Sentry. The system ([android 8+](https://developer.android.com/about/versions/oreo/background#services)) prevents apps from starting services while they are in the background without showing a notification so the user is aware of what is going on. If an app invokes `startService` from the background an IllegalStateException is thrown. Apps can call `startForegroundService` instead, which gives them 5s to show a notification.


This PR makes sure `UploadService` is started in foreground so the IllegalStateException is not thrown.
- The non-fatal exception was happening under the following circumstances. If `LocalDraftUploadStarter` was started from a foreground but before `uploadServiceFacade.uploadPost` was called the app went to background, when we called `startService` the `IllegalStateException` was thrown. The exception was caught [here](https://github.com/wordpress-mobile/WordPress-Android/blob/ac56c495d4f5cecf2a50f349576f23515679d8b1/WordPress/src/main/java/org/wordpress/android/ui/uploads/LocalDraftUploadStarter.kt#L94). 
- I decided to start the service as a foreground service as I think it's ok to start the upload manually (without the use of a JobScheduler/WorkerManager) when the app just went to background. The `UploadWorker` takes care of uploading posts when the app has been in the background for a while.

Note:
~We are not showing any notification when we are retrying upload at the moment  - however, this issue should be fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/9956.~

To test:
1. Upload a post with and without media

- I don't think it's necessary to test whether the app doesn't throw the IllegalStateException. We can check Sentry when this version is in production as it's a non-fatal exception. We could somehow add an artificial delay and test it, but I don't think it's worth it.


Update release notes:

- [ x ] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.
